### PR TITLE
Migrate multisig

### LINKED ISSUE #217: Process multisig state

### PR DESCRIPTION
**Pull request by AurevoirXavier**
Closes #217.

### CODE REVIEW

File: pallet/account-migration/src/lib.rs [hunk 1]
```diff
@@ -173,8 +185,157 @@ pub mod pallet {
...
+			} else {
+				let mut members = members.into_iter().map(|m| (m, false)).collect::<Vec<_>>();
+
+				// Set the status to `true`.
+				//
+				// Because the `_signature` was already been verified in `pre_dispatch`.
+				members
+					.last_mut()
+					.expect("[pallet::account-migration] `members` will never be empty; qed")
+					.1 = true;
```
Comment (boundless-forest) [inline pallet/account-migration/src/lib.rs:221]: Do we need to `threshold + 1`  after updating the member's status?

Comment (AurevoirXavier) [reply pallet/account-migration/src/lib.rs:221]: Threshold is a fixed value.

e.g. `3/5` is always `3/5`.

File: pallet/account-migration/src/lib.rs [hunk 2]
```diff
@@ -173,8 +185,157 @@ pub mod pallet {
...
+			if threshold < 2 {
+				Self::migrate_inner(multisig, to)?;
+			} else {
+				let mut members = members.into_iter().map(|m| (m, false)).collect::<Vec<_>>();
+
+				// Set the status to `true`.
+				//
+				// Because the `_signature` was already been verified in `pre_dispatch`.
+				members
+					.last_mut()
```
Comment (boundless-forest) [inline pallet/account-migration/src/lib.rs:219]: I guess you want to get `who` from members by using `last_mut()`?

If so, I think it is incorrect. The members returned from `multisig_of(xxx)` is sorted.

Comment (AurevoirXavier) [reply pallet/account-migration/src/lib.rs]: Yep. Because I refactored some of them.

File: pallet/account-migration/src/lib.rs [hunk 3]
```diff
@@ -173,8 +185,158 @@ pub mod pallet {
...
+					Self::pre_check_signature(who, to, signature)
+				},
+				Call::complete_multisig_migration { multisig, submitter, signature } => {
+					let Some(multisig_info) = <Multisigs<T>>::get(multisig) else {
+						return InvalidTransaction::Custom(E_ACCOUNT_NOT_FOUND).into();
+					};
+
+					if !multisig_info.members.iter().any(|(a, _)| a == submitter) {
+						return InvalidTransaction::Custom(E_NOT_MULTISIG_MEMBER).into();
+					}
```
Comment (boundless-forest) [inline pallet/account-migration/src/lib.rs:292]: I think we should reject useless requests from members whose status is already `true`.

**Comment by hackfisher** [comment]
If the `to` for multisig migration is also a new multisig, does it pre-exist? Will it violate the checking of non-existing requirement?

Another thing I'm checking is whether the singature used for multisig operator can be reuse/replayed in the multisig case, but it seems that the `to` info in message will prevent that, since no two operators self `to` can be the same?

File: pallet/account-migration/src/lib.rs [hunk 4]
```diff
@@ -173,8 +185,158 @@ pub mod pallet {
...
+			} else {
+				let mut members = members.into_iter().map(|m| (m, false)).collect::<Vec<_>>();
+
+				// Set the status to `true`.
+				//
+				// Because the `_signature` was already been verified in `pre_dispatch`.
+				members
+					.iter_mut()
+					.find(|(a, _)| a == &who)
+					.expect("[pallet::account-migration] `members` will never be empty; qed")
```
Comment (boundless-forest) [inline pallet/account-migration/src/lib.rs:221]: Please update the expected info string, it's stale.

**Review by hackfisher — APPROVED** [review]

**Review by boundless-forest — APPROVED** [review]

**Pull request merged**